controller/src/main/java/com/epam/esm/controller/GiftCertificateController.java

```diff
@@ -0,0 +1,128 @@
+package com.epam.esm.controller;
+
+import com.epam.esm.model.dao.entity.SortType;
+import com.epam.esm.model.service.GiftCertificateService;
+import com.epam.esm.model.service.dto.CertificateDTO;
+import com.epam.esm.model.service.exception.ServiceException;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+
+/**
+ * Gift certificate RestAPI.
+ */
+@RestController
+@RequestMapping("/certificate")
+public class GiftCertificateController {
+
+    private final GiftCertificateService certificateService;
+
+    private GiftCertificateController(GiftCertificateService certificateService) {
+        this.certificateService = certificateService;
+    }
+
+    /**
+     * Find certificate by id.
+     *
+     * @param id the id
+     * @return the response entity
+     * @throws ServiceException the service exception
+     */
+    @GetMapping("/{id}")
+    public ResponseEntity<CertificateDTO> find(@PathVariable(name = "id") Long id) throws ServiceException {
+        CertificateDTO certificateDTO = certificateService.find(id);
+        return ResponseEntity.ok(certificateDTO);
+    }
+
+    /**
+     * Add certificate.
+     *
+     * @param certificateDTO the certificate dto
+     * @return the response entity
+     * @throws ServiceException the service exception
+     */
+    @PostMapping(value = "/")
+    public ResponseEntity<CertificateDTO> add(@RequestBody CertificateDTO certificateDTO) throws ServiceException {
+        CertificateDTO result = certificateService.add(certificateDTO);
+        return ResponseEntity.ok(result);
+    }
+
+    /**
+     * Update certificate. Mark the fields that are not specified for updating null.
+     *
+     * @param certificateDTO the certificate dto
+     * @return the response entity
+     * @throws ServiceException the service exception
+     */
+    @PutMapping(value = "/", consumes = "application/json")
+    public ResponseEntity<CertificateDTO> update(@RequestBody CertificateDTO certificateDTO) throws ServiceException {
+        CertificateDTO result = certificateService.update(certificateDTO);
+        return ResponseEntity.ok(result);
+    }
+
+    /**
+     * Delete certificate by id.
+     *
+     * @param id the id
+     * @return the response entity
+     * @throws ServiceException the service exception
+     */
+    @DeleteMapping("/{id}")
+    public ResponseEntity<String> delete(@PathVariable(name = "id") Long id) throws ServiceException {
+        long result = certificateService.delete(id);
+        return result != -1L ? ResponseEntity.ok("Delete successful!") : ResponseEntity.ok("Delete unsuccessful!");
+    }
+
+    /**
+     * Find certificate by tag name.
+     *
+     * @param tagName the tag name
+     * @return the response entity
+     * @throws ServiceException the service exception
+     */
+    @GetMapping("/byName/{name}")
```

Comment:
Why did you decide to split one operation with finding by different parameters to different operations with finding by single parameter? I guess this is not up to the task

Reply:
Moreover, If you decide to leave it as it is, this naming is not suitable, please read more about naming practices of rest API

Reply:
The assignment said to sort by date or name. 
Sorting by both tasks was extra task

Reply:
I mean that all methods for searching and sorting that you have implemented should not be separate, they should be parameters of one request, so API should have possibility to find certificates by all params together

Reply:
I redid it again, hope now I understand correctly